### PR TITLE
VDiff: Support diffing tables without a defined Primary Key

### PR DESCRIPTION
**Pull request by mattlord**
## Description

VReplication supports tables without defined Primary Keys, but VDiff did not. If you executed a VDiff on a workflow that included a table without one, it would fail like this (see [issue](https://github.com/vitessio/vitess/issues/14793)):
```
VDiff Summary for customer.commerce2customer (4ecb6a28-9263-404e-a4e9-d737ecbca5e5)
State:        error
              Error: (shard 0) buildPlan: error getting PK column collations for table tnopk: empty list supplied for vars2
RowsCompared: 0
HasMismatch:  false
StartedAt:    2023-12-15 21:29:47

Use "--format=json" for more detailed output
```

This PR adds support for diffing tables that have no defined Primary Key. It uses a Primary Key equivalent (unique index on non-null column(s)) if one exists, and if not then all of the columns in the table as a substitute — just as does [VReplication](https://github.com/vitessio/vitess/blob/7c9fec7205cb27e34ae3db6b365896120aa918fb/go/vt/vttablet/tabletmanager/vreplication/vreplicator.go#L347-L360), and as does [MySQL row-based replication](https://dev.mysql.com/doc/mysql-replication-excerpt/8.0/en/replication-features-row-searches.html) (although there it will use any index if there is one first). 

Basic manual test:
```
./101_initial_cluster.sh

mysql < ../common/insert_commerce_data.sql

./201_customer_tablets.sh

mysql commerce -e "create table nopk (name varchar(100), age bigint unsigned)"

for _ in {1..15}; do
  mysql commerce -e "insert into nopk values ('${RANDOM}_person_${RANDOM}', ${RANDOM}); insert into nopk select * from nopk;"
done

vtctldclient MoveTables --workflow commerce2customer --target-keyspace customer create --source-keyspace commerce --all-tables

vtctldclient vdiff --workflow=commerce2customer --target-keyspace=customer create

sleep 10

vtctldclient vdiff --workflow=commerce2customer --target-keyspace=customer show last --verbose
```

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/14793

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added: https://github.com/vitessio/website/pull/1651

### CODE REVIEW

File: go/vt/mysqlctl/schema.go [hunk 1]
```diff
@@ -579,13 +579,7 @@ func (mysqld *Mysqld) ApplySchemaChange(ctx context.Context, dbName string, chan
 // defined PRIMARY KEY then it may return the columns for
 // that index if it is likely the most efficient one amongst
 // the available PKE indexes on the table.
-func (mysqld *Mysqld) GetPrimaryKeyEquivalentColumns(ctx context.Context, dbName, table string) ([]string, string, error) {
-	conn, err := getPoolReconnect(ctx, mysqld.dbaPool)
```
Comment (rohit-nayak-ps) [inline go/vt/mysqlctl/schema.go:583]: Did the functionality change for this PR require a change in how we acquire a connection here?

**Review by rohit-nayak-ps — APPROVED** [review]

File: go/vt/mysqlctl/schema.go (hunk 1 above)
Comment (mattlord) [reply go/vt/mysqlctl/schema.go:583]: When I first created this I had a `mysqlctl.Mysqld` instance readily available in the callsite. In working on this, I realized that I should have changed it when using it in vstreamer as I created a new instance there, which is heavy as it creates conn pools etc (and I wasn't properly calling close in a defer 🤦‍♂️), so I changed it here to use a callback to talk to the DB instead. It's much lighter and is far better (and is an already established pattern in the mysqlctl package).

Comment (shlomi-noach) [reply go/vt/mysqlctl/schema.go:583]: If we adopt this pattern, then we will find ourselves passing `exec` functions in a large amount of functions - not saying it's wrong - just thinking of the code/signature impact.

File: go/vt/vttablet/tabletmanager/vdiff/table_plan.go [hunk 2]
```diff
@@ -153,10 +155,25 @@ func (td *tableDiffer) buildTablePlan(dbClient binlogplayer.DBClient, dbName str
...
+		// We use the columns from a PKE if there is one.
+		pkeCols, err := tp.getPKEquivalentColumns(dbClient)
+		if err != nil {
+			return nil, vterrors.Wrapf(err, "error getting PK equivalent columns for table %s", tp.table.Name)
+		}
+		if len(pkeCols) > 0 {
+			tp.table.PrimaryKeyColumns = append(tp.table.PrimaryKeyColumns, pkeCols...)
+		} else {
+			// We use every column together as a substitute PK.
+			tp.table.PrimaryKeyColumns = append(tp.table.PrimaryKeyColumns, tp.table.Columns...)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/vdiff/table_plan.go:168]: Does it make sense to allow full scan in VDiff? Should we instead say this table isn't supported? I'm imagining a VDiff running over weeks - as a user of VDiff I'd not want it to run so long.

**Review by shlomi-noach — APPROVED** [review]
Looks good with one question about allowing full table scans.

Also, it looks like we're implementing similar behaviors in different approaches; the vstreamer/vplayer approach uses different code paths, right?

File: go/vt/vttablet/tabletmanager/vdiff/table_plan.go (hunk 2 above)
Comment (mattlord) [reply go/vt/vttablet/tabletmanager/vdiff/table_plan.go:168]: IMO, yes. We allow it in VReplication as well. This is up to the user -- it's not uncommon to have small tables w/o a primary key or PKE. In some cases all queries against the table will be a full scan. It's not our call to say that you shouldn't be able to move those tables and diff them after doing so.

File: go/vt/mysqlctl/schema.go (hunk 1 above)
Comment (mattlord) [reply go/vt/mysqlctl/schema.go:583]: It's already a pattern (in that file/package and elsewhere). I agree that it's not one which should be used w/o good reason.

**Comment by mattlord** [comment]
> Also, it looks like we're implementing similar behaviors in different approaches; the vstreamer/vplayer approach uses different code paths, right?

I'm not sure what you're referring to here. The vreplicator, vstreamer, and now vdiff are all using the same function and code path for managing PKEs. Maybe I'm misunderstanding?

**Comment by shlomi-noach** [comment]
> I'm not sure what you're referring to here. The vreplicator, vstreamer, and now vdiff are all using the same function and code path for managing PKEs. Maybe I'm misunderstanding?

No, you were good. I wasn't clear, even to myself. OK, the example I was thinking of was actually Online DDL, where we can use ay non-NULLable unique key for the migration. Context: https://github.com/vitessio/vitess/pull/8364.

In Online DDL, we get the shared unique key of two tables:

https://github.com/vitessio/vitess/pull/8364/files#diff-552cf284bd2c1a9e4ae1c9f9f1c350c7dcc04177ed3a663b0670d084518e7fe1R107

Having read a table's unique keys:

https://github.com/vitessio/vitess/pull/8364/files#diff-552cf284bd2c1a9e4ae1c9f9f1c350c7dcc04177ed3a663b0670d084518e7fe1R205

We populate the Rule with unique key info:

https://github.com/vitessio/vitess/pull/8364/files#diff-5e69c397e15dcfda72bb6fe8dba93b2d11a8d2d7b594ebfbbdcfdc07bfc3791dR165

We then read that as needed to override PK columns:

https://github.com/vitessio/vitess/pull/8364/files#diff-944642971422bb00ed68466f784f8b586f5e800f57e0c46002d6194c517cb1a0R526-R533

**Pull request merged**